config/include.go

```diff
@@ -35,7 +35,23 @@ func parseIncludedConfig(
 		includePath = util.JoinPath(filepath.Dir(terragruntOptions.TerragruntConfigPath), includePath)
 	}
 
-	if len(decodeList) > 0 {
+	// These condition are here to specifically handle the `run-all apply` command on a new deployment. During
+	// `run-all apply`, terragrunt needs to first build up the dependency graph to know what order to apply the modules
+	// in. For a completely new deployment, no outputs are available. This means that we can't parse anything that
+	// relies on a `dependency` during this stage, as the outputs are not materialized until later in the process when
+	// terragrunt actually starts applying the modules. To support this, we implement the following conditions for when
+	// terragrunt can fully parse the included config (only one needs to be true):
+	// - Included config does NOT have a dependency block.
+	// - Terragrunt is NOT performing a partial parse (which indicates whether or not Terragrunt is building a module
+	//   graph).
+	// To make the logic easier to implement, we implement the inverse here, where we check whether the included config
+	// has a dependency block, and if we are in the middle of a partial parse, we perform a partial parse of the
+	// included config.
```

Comment:
A few thoughts:

1. Doesn't this impact _all_ `run-all` commands? That is, not just `run-all apply`, and not even just on the first run? For all these commands, Terragrunt needs to build a dependency graph; to do that, it has to find all the `dependency` and `dependencies` blocks; if any of those depend on an output of another `dependency`, then we won't know the value of that output until _after_ running `apply`. There are no outputs at all if you've never run `run-all apply`, but there may also not be outputs if this is a totally new dependency that has never been applied (so you did `run-all apply` on _some_ modules, but not all), or, even if `apply` had been run before, the outputs may be different now, and we won't know that without running `apply` again. E.g., I may have a `dependency` on some module `foo`, where I added a new output `bar`, but haven't run `apply` yet; if I have a `dependency` block with the path set to `dependency.foo.outputs.bar`, then it won't work until after `apply`. So how do these dynamic `dependency` paths really work in practice?
1. NIT on the comment contents themselves: it needs a tiny clarification right after the second sentence where we explain that we build a dependency graph by looking for `dependency` and `dependencies` blocks, which works fine when those are all defined statically, but gets tricky if the paths in those blocks depend on dynamic outputs of other `dependency` blocks.

Reply:
Tried another take to expand on the comment here. Hopefully this nails all the nuances, what is happening here, and why we are doing it. 568b631a1ac533e36856323b911fd56760c08e7c